### ISSUE TITLE
gRPC Ingress with wildcard host match returns "Unimplemented" on client side 

### ISSUE THREAD

**Issue by mkurde** [istio/istio#27320]
**Bug description**

After upgrading to Istio 1.7.1 (from 1.5.4) the authority header matching in ingress-grpc VirtualServices stopped working and clients get "Unimplemented". 

Not working  (Working on Istio 1.5.4)
```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: ingress-grpc
  namespace: default
spec:
  gateways:
  - shared/istio-ingress
  hosts:
  - '*.default.foo.bar'
  http:
  - match:
    - port: 6565
      authority:
        prefix: 'bob.default'
    route:
    - destination:
        host: bob.default.svc.cluster.local
        port:
          number: 6565
  - match:
    - port: 6565
      authority:
        prefix: 'alice.default'
    route:
    - destination:
        host: alice.default.svc.cluster.local
        port:
          number: 6565
```

We could get it working by splitting up the entries into single VirtualSerive Resources without authority header matching.

Working 👍 
```
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: ingress-grpc-bob
  namespace: default
spec:
  gateways:
  - shared/istio-ingress
  hosts:
  - bob.default.foo.bar
  http:
  - match:
    - port: 6565
    route:
    - destination:
        host: bob.default.svc.cluster.local
        port:
          number: 6565
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: ingress-grpc-alice
  namespace: default
spec:
  gateways:
  - shared/istio-ingress
  hosts:
  - alice.default.foo.bar
  http:
  - match:
    - port: 6565
    route:
    - destination:
        host: alice.default.svc.cluster.local
        port:
          number: 6565
```

Gateway configuration used in the example
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-ingress
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 6565
      name: grpc
      protocol: grpc
    hosts:
    - "*"
  - port:
      number: 80
      name: http2
      protocol: http2
    hosts:
    - "*" 
```

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
One gRPC VirtualService with multiple matching prefixes on a host with leading wildcard.

**Steps to reproduce the bug**

Apply two gRPC services (alice and bob) and the mentioned Gateway "istio-ingress". 

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**

```
→ istioctl version --remote
client version: 1.7.0
control plane version: 1.7.1
data plane version: 1.7.1 (91 proxies)

→ kubectl version --short
Client Version: v1.19.0
Server Version: v1.17.9-gke.600
```

**How was Istio installed?**
istio operator

**Environment where bug was observed (cloud vendor, OS, etc)**
* GKE
* Linux

**Comment by arnecls** [comment]
We could not find a reason why exactly the wildcard host is causing issues here, yet.
All envoy configurations look OK (checked via `istioctl pc`)

It does not seem to be an issue with HTTP/HTTP2 services, which use the exact same pattern.
I.e. HTTP2 services listening on port 80 still use wildcard host + authority match with istio 1.7.1, as they did with istio 1.5.4.

**Comment by howardjohn** [comment]
Might be the same issue as https://github.com/istio/istio/issues/25350? Easy way to check would be to look at the route config dump for pattern like `  "*.domain.net:*"`

**Comment by arnecls** [comment]
Looks like it.
The pattern you mentioned is there.
We've spotted the additional ports via Wireshark, too.

**Comment by arnecls** [comment]
This can be closed as a duplicate of #25350.
Corresponding envoy issues is https://github.com/envoyproxy/envoy/issues/12647

**Comment by howardjohn** [comment]
@arnecls I think when we fixed some cases in some version after 1.5.4, we broke this one. I think we can make both cases better, but the "perfect" solution will require envoy change to support wildcard domain on arbitrary port (ie support nodeport or other fronted LBs where port does not match)

**Comment by howardjohn** [comment]
Will track in https://github.com/istio/istio/issues/25350
